### PR TITLE
Refactor Ensemble Functions & Change Ensemble APIs

### PR DESCRIPTION
**Pull request by michaelmckinsey1**
# Summary
`thicket.py` is becoming quite unwieldy so this refactor is in an effort to improve readability, maintainability and usability.

1. API name changes. Old API now errors.
   - `Thicket.columnar_join` -> `Thicket.concat_thickets(axis="columns")`
   - `Thicket.unify_ensemble` -> `Thicket.concat_thickets(axis="index")`
2. 3 separate unify functions now made into 1 function, `_unify`.
3. Functions are now hidden away from the user and in Thicket we only expose the functions we want to be user facing.
4. Refactoring of sub-functions.

We should consider doing something similar for the filter functions, etc.

### CODE REVIEW

**Review by cscully-allison — APPROVED** [review]
PR looks pretty good, just one small comment.

File: thicket/ensemble.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+            for k, v in profile_mapping_cp.items():
+                combined_th.profile_mapping[
+                    new_mappings[k]
+                ] = combined_th.profile_mapping.pop(k)
+            combined_th.performance_cols = helpers._get_perf_columns(
+                combined_th.dataframe
+            )
+
+        def _handle_perfdata():
+            """Handle performance data table operations.
```
Comment (cscully-allison) [inline thicket/ensemble.py]: This doc string and function name could describe the functionality a little more.

Comment (michaelmckinsey1) [reply thicket/ensemble.py]: [f18b8fa](https://github.com/LLNL/thicket/pull/61/commits/f18b8faa73455af8c6776a1bb22f2e1f6801bd83)

**Comment by slabasan** [comment]
@vanessalama09 This PR is making changes to the tutorial notebooks, but to the rendered ones in the thicket source. How would you like to get these changes over into the thicket-docs version?

File: docs/thicket_tutorial.ipynb
Comment (slabasan) [inline docs/thicket_tutorial.ipynb:1]: Changes should be in thicket-tutorial, and then rendered here

File: docs/thicket_rajaperf_clustering.ipynb
Comment (slabasan) [inline docs/thicket_rajaperf_clustering.ipynb:1]: Changes to notebooks should be in thicket-tutorial, then rendered here

File: docs/thicket_tutorial.ipynb
Comment (michaelmckinsey1) [reply docs/thicket_tutorial.ipynb:1]: [440a036](https://github.com/LLNL/thicket/pull/61/commits/440a036de87cc34063e8259f25cffc5d79ebc557)

File: docs/thicket_rajaperf_clustering.ipynb
Comment (michaelmckinsey1) [reply docs/thicket_rajaperf_clustering.ipynb:1]: [440a036](https://github.com/LLNL/thicket/pull/61/commits/440a036de87cc34063e8259f25cffc5d79ebc557)

**Review by slabasan — APPROVED** [review]

**Comment by slabasan** [comment]
@MichaelMcKinsey1 Can you please rebase this branch on top of develop to fix the conflict?

**Comment by vanessalama09** [comment]
> @vanessalama09 This PR is making changes to the tutorial notebooks, but to the rendered ones in the thicket source. How would you like to get these changes over into the thicket-docs version?

Thicket-docs are in thicket source. The rendered versions are the ones in docs. Did you mean over to thicket-tutorial?

**Review by slabasan — APPROVED** [review]

**Pull request merged**